### PR TITLE
Deep Speed compatibility + checkpoint saving bug fixes

### PR DESCRIPTION
**Pull request by mihirneal**
# What does this PR do?

Added compatibility with DeepSpeed which includes adding a dummy learning rate scheduler and optimizer when use_deepspeed argument is set to True (default False).

Issues fixed:

- Validation generated images now can be viewed on WandB (or any other platform). Earlier it used to blacked out. Probably wasn't generating images properly.
- Checkpoint saving now works for multi-GPU training. Earlier it was in a deadlock since accelerator only saved those checkpoints if it was the main process and then invoke accelerator.wait_for_everyone which would put it in an infinite loop.
- Readme for Stable Diffusion XL Control Net mentions the updated, more stable SDXL VAE. The original one absolutely doesn't work.

### CODE REVIEW

**Comment by mihirneal** [comment]
Can't assign people to this PR it seems so just tagging y'all here: @sayakpaul @patrickvonplaten @pacman100

**Comment by mihirneal** [comment]
This is regarding training Control Net for Stable Diffusion XL. Forgot to mention that.

File: examples/controlnet/train_controlnet_sdxl.py [hunk 1]
```diff
@@ -80,9 +80,6 @@ def log_validation(vae, unet, controlnet, args, accelerator, weight_dtype, step)
     pipeline = pipeline.to(accelerator.device)
     pipeline.set_progress_bar_config(disable=True)
 
-    if args.enable_xformers_memory_efficient_attention:
```
[the remark this answers is not in the document]
Comment (mihirneal) [reply examples/controlnet/train_controlnet_sdxl.py:83]: Used to have issues with it, but seems like it works now. I've added it back.

**Comment by bghira** [comment]
if wait-for-everyone causes endless loop, chances are you need apply_dataset_padding used

**Comment by bghira** [comment]
can you revert the changes about selecting a vae? additionally if we add a default negative prompt it should be a cmdline argument

**Comment by mihirneal** [comment]
I'll add default negative prompts as cmd arguments, however the original SDXL VAE does not work (at least not in my tests), a lot of NaNs and the image is basically all noise. Using this vae solves the issue.

**Comment by mihirneal** [comment]
> if wait-for-everyone causes endless loop, chances are you need apply_dataset_padding used

could you lemme know the specifics about this? Removing wait_for_everyone solves the checkpointing issue since it's handled directly by deepspeed / accelerate (saving states for each GPU in case of multi GPU training). Why would i need to add additional code.

**Comment by bghira** [comment]
because for instances where we're not using deepspeed, the GPUs run in separate processes, and we should not be removing that for those situations. they will all unnecessarily run validations together, and then upload to WandB.

if this is a DeepSpeed specific issue, i think the accelerator.state has a component that tells us whether deepspeed is in use, which then allows:

```py
if not accelerator.state.deep_speed:
    accelerator.wait_for_everyone
```

**Comment by bghira** [comment]
setting the default vae path argument value to the madebyollin VAE would be awesome, that is how i have handled the same issue. then, a user can explicitly override it with the vanilla 1.0 or 0.9 VAE, should they so choose.

**Comment by bghira** [comment]
we can use:

```py
is_ds_zero_3 = False
if getattr(accelerator.state, "deepspeed_plugin", None):
    is_ds_zero_3 = accelerator.state.deepspeed_plugin.zero_stage == 3
```

instead of requiring the user to pass `--use_deepspeed` which should make this nicely seamless

File: examples/controlnet/train_controlnet_sdxl.py [hunk 2]
```diff
@@ -103,18 +102,22 @@ def log_validation(vae, unet, controlnet, args, accelerator, weight_dtype, step)
         )
 
     image_logs = []
-
+    negative_prompt = "anime, low quality, bad quality, sketches, cartoon, graphic, text, painting, crayon, graphite, abstract, glitch, deformed, mutated, ugly, disfigured"
```
Comment (patrickvonplaten) [inline examples/controlnet/train_controlnet_sdxl.py:105]: I don't think this is related to the PR no? Could we maybe remove this extra line?

File: examples/t2i_adapter/train_t2i_adapter_sdxl.py [hunk 3]
```diff
@@ -78,21 +79,22 @@ def log_validation(vae, unet, adapter, args, accelerator, weight_dtype, step):
...
     adapter = accelerator.unwrap_model(adapter)
-
+    euler_a = EulerAncestralDiscreteScheduler.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="scheduler"
+    )
     pipeline = StableDiffusionXLAdapterPipeline.from_pretrained(
         args.pretrained_model_name_or_path,
         vae=vae,
-        unet=unet,
+        scheduler=euler_a,
```
Comment (bghira) [inline examples/t2i_adapter/train_t2i_adapter_sdxl.py:88]: huh?

File: examples/t2i_adapter/train_t2i_adapter_sdxl.py [hunk 4]
```diff
@@ -78,21 +79,22 @@ def log_validation(vae, unet, adapter, args, accelerator, weight_dtype, step):
...
         adapter=adapter,
         revision=args.revision,
         torch_dtype=weight_dtype,
+        variant="fp16",
```
Comment (bghira) [inline examples/t2i_adapter/train_t2i_adapter_sdxl.py:92]: hardcoding this isnt good

File: examples/t2i_adapter/train_t2i_adapter_sdxl.py [hunk 5]
```diff
@@ -113,18 +115,23 @@ def log_validation(vae, unet, adapter, args, accelerator, weight_dtype, step):
...
+    negative_prompt = "anime, low quality, bad quality, sketches, cartoon, graphic, text, painting, crayon, graphite, abstract, glitch, deformed, mutated, ugly, disfigured"
     for validation_prompt, validation_image in zip(validation_prompts, validation_images):
         validation_image = Image.open(validation_image).convert("RGB")
         validation_image = validation_image.resize((args.resolution, args.resolution))
 
         images = []
 
         for _ in range(args.num_validation_images):
-            with torch.autocast("cuda"):
-                image = pipeline(
```
Comment (bghira) [inline examples/t2i_adapter/train_t2i_adapter_sdxl.py:125]: we need autocast for bf16

File: examples/t2i_adapter/train_t2i_adapter_sdxl.py [hunk 6]
```diff
@@ -113,18 +115,23 @@ def log_validation(vae, unet, adapter, args, accelerator, weight_dtype, step):
...
-                    prompt=validation_prompt, image=validation_image, num_inference_steps=20, generator=generator
-                ).images[0]
+            image = pipeline(
+                prompt=validation_prompt,
+                negative_prompt=negative_prompt,
+                image=validation_image,
+                num_inference_steps=30,
+                adapter_conditioning_scale=0.8,
```
Comment (bghira) [inline examples/t2i_adapter/train_t2i_adapter_sdxl.py:131]: can the new values have parameters instead of hard-coding

**Comment by bghira** [comment]
i've been messing with DeepSpeed and it seems that there's potentially a few more code changes required, than this.

the code you have proposed, only works with a "scheduler" and "optimizer" defined in the deepspeed config yaml file.

the dummyoptim and dummyscheduler class will throw an error if you try and load them when no scheduler/optimizer is defined in the DeepSpeed config.

see:

```py
    use_deepspeed_optimizer = False
    use_deepspeed_scheduler = False
    if (
        hasattr(accelerator.state, "deepspeed_plugin")
        and accelerator.state.deepspeed_plugin is not None
    ):
        offload_param = accelerator.state.deepspeed_plugin.deepspeed_config[
            "zero_optimization"
        ]["offload_param"]
        accelerator.state.deepspeed_plugin.deepspeed_config["zero_optimization"][
            "offload_param"
        ]["pin_memory"] = False
        if offload_param["device"] == "nvme":
            if offload_param["nvme_path"] == "none":
                if args.offload_param_path is None:
                    raise ValueError(
                        f"DeepSpeed is using {offload_param['device']} but nvme_path is not specified."
                    )
                else:
                    accelerator.state.deepspeed_plugin.deepspeed_config[
                        "zero_optimization"
                    ]["offload_param"]["nvme_path"] = args.offload_param_path
            logger.info(
                f"Using DeepSpeed NVMe offload at {accelerator.state.deepspeed_plugin.deepspeed_config['zero_optimization']['offload_param']['nvme_path']}."
            )

        use_deepspeed_optimizer = True
        if "optimizer" not in accelerator.state.deepspeed_plugin.deepspeed_config:
            accelerator.state.deepspeed_plugin.deepspeed_config["optimizer"] = {
                "type": "AdamW",
                "params": {
                    "lr": args.learning_rate,
                    "betas": [args.adam_beta1, args.adam_beta2],
                    "eps": args.adam_epsilon,
                    "weight_decay": args.adam_weight_decay,
                },
            }

        use_deepspeed_scheduler = True
        if "scheduler" not in accelerator.state.deepspeed_plugin.deepspeed_config:
            accelerator.state.deepspeed_plugin.deepspeed_config["scheduler"] = {
                "type": "WarmupLR",
                "params": {
                    "warmup_min_lr": 0,
                    "warmup_max_lr": args.learning_rate,
                    "warmup_num_steps": args.lr_warmup_steps,
                },
            }
```

additionally, the default deepspeed ZeRO 3 config doesn't indicate correctly where to store the NVMe offload files, and it's possible to do so from within the code, using parse_args parameters.

**Comment by bghira** [comment]
```
AttributeError: 'DummyOptim' object has no attribute 'step'
```

this happens when using the DeepSpeed-defined optimizer/scheduler config. how did you handle this? it seems like your code directly calls `step()` on the dummy optimizer still.

**Pull request closed without merging**